lang/eval.go

```diff
@@ -240,15 +240,17 @@ func (s *Scope) evalContext(refs []*addrs.Reference, selfAddr addrs.Referenceabl
 			// Self is an exception in that it must always resolve to a
 			// particular instance. We will still insert the full resource into
 			// the context below.
+			var hclDiags hcl.Diagnostics
 			switch k := subj.Key.(type) {
 			case addrs.IntKey:
-				self = val.Index(cty.NumberIntVal(int64(k)))
+				self, hclDiags = hcl.Index(val, cty.NumberIntVal(int64(k)), ref.SourceRange.ToHCL().Ptr())
+				diags.Append(hclDiags)
 			case addrs.StringKey:
-				self = val.Index(cty.StringVal(string(k)))
+				self, hclDiags = hcl.Index(val, cty.StringVal(string(k)), ref.SourceRange.ToHCL().Ptr())
+				diags.Append(hclDiags)
```

Comment:
I'm pondering what happens if `hcl.Index` fails here.

It looks like it shouldn't happen because `val` should always be of a suitable type by the time we get here (or else we would've been panicking before anyway), so maybe not a big deal, but perhaps worth adding a comment to that effect anyway to assuage the concerns of future readers.

It looks like `hcl.Index` is consistent with the usual HCL convention that if it returns errors then it also returns an incomplete-but-valid placeholder value (`cty.DynamicVal`, in this case) so later code below will hopefully be okay with `self` being `cty.DynamicVal` and just short-circuit as normal and allow the error to be returned. Might therefore be okay to just add a comment here noting that we're not expecting this to return errors, but if it does then `self` will be `cty.DynamicVal`, in case that's helpful to a future maintainer.

Reply:
That matches my thoughts here too. I'm actually not sure which is better now, since in the rare case of an internal error making it this far we `panic` with the exact location. I'm still leaning towards collecting the error and passing it along for user ergonomics, so I'll just add another comment for now.